farm/modeling/prediction_head.py

```diff
@@ -849,66 +881,174 @@ def prepare_labels(self, start_position, end_position, **kwargs):
+                    if self.top_n_predictions > 1:
+                        passage_pred["prediction_rank"] = i
+                    s_i = current_pred[i,0]
+                    e_i = current_pred[i,1]
+                    logit_sum_i = current_pred[i,2]
+                    question_shift_i = current_pred[i,4]
+                    passage_shift_i = current_pred[i,5]
+                    passage_pred["logit_sum"] = logit_sum_i
+                    try:
+                        #default to returning no answer
+                        start = 0
+                        end = 0
+                        context_start = 0
+                        answer = ""
+                        context = ""
+                        if(s_i + e_i > 0):
+                            current_start = int(s_i + passage_shift_i - question_shift_i)
+                            current_end = int(e_i + passage_shift_i - question_shift_i) + 1
+                            start = current_sample.tokenized["offsets"][current_start]
+                            end = current_sample.tokenized["offsets"][current_end]
+                            # we want the answer in original string space (containing newline, tab or multiple
+                            # whitespace. So we need to join doc tokens and work with character offsets
+                            temptext = " ".join(current_sample.clear_text["doc_tokens"])
+                            answer = temptext[start:end]
+                            context_start = int(np.clip((start-self.context_size),a_min=0,a_max=None))
+                            context_end = int(np.clip(end +self.context_size,a_max=len(temptext),a_min=None))
+                            context = temptext[context_start:context_end]
+                            answer = answer.strip()
+                    except IndexError as e:
+                        logger.info(e)
+                    passage_pred["answer"] = answer
+                    passage_pred["context"] = context
+                    passage_pred["offset_start"] = start - context_start
+                    passage_pred["offset_end"] = end - context_start
+                    passage_pred["document_id"] = current_sample.clear_text.get("document_id", None)
+                    passage_predictions.append(passage_pred)
+
+                pred = {}
+                pred["question"] = current_sample.clear_text.question_text
+                pred["question_id"] = current_sample.clear_text.get("qas_id", None)
+                pred["ground_truth"] = current_sample.clear_text.get("orig_answer_text", None)
+                pred["answers"] = passage_predictions
+                all_preds.append(pred)
 
         result["predictions"] = all_preds
         return result
 
-    def _get_best_indexes(self, logits, n_best_size):
-        """Get the n-best logits from a numpy array."""
-        idx = np.argsort(logits,axis=1)[:,-n_best_size:]
+    def _aggregate_predictions(self, preds):
```

Comment:
Can we rename this to "aggregate_preds" to be more compliant with the rest (formatted_preds, logits_to_preds...). I would also make this a public method, since there might be cases (e.g. in Branden's branch in eval) where we call this method from the outside.